flink-cdc-runtime/src/main/java/com/ververica/cdc/runtime/operators/schema/encoding/SchemaEncodingFunction.java

```diff
@@ -0,0 +1,273 @@
+
+/**
+ * Function for encoding {@link Schema} into {@link DataChangeEvent} and generate {@link
+ * SchemaAwareDataChangeEvent}.
+ */
+public class SchemaEncodingFunction extends RichMapFunction<Event, Event> {
+    private final MapStateDescriptor<TableId, Schema> schemaStateDescriptor =
+            new MapStateDescriptor<>("schemas", TableId.class, Schema.class);
+    private SchemaHolder schemaHolder;
+
+    @Override
+    public void open(Configuration parameters) throws Exception {
+        schemaHolder = new SchemaHolder(getRuntimeContext().getMapState(schemaStateDescriptor));
+        super.open(parameters);
+    }
+
+    @Override
+    public Event map(Event event) throws Exception {
+        if (event instanceof SchemaChangeEvent) {
+            schemaHolder.applySchemaChange(((SchemaChangeEvent) event));
+            return event;
+        } else if (event instanceof DataChangeEvent) {
+            DataChangeEvent dataChangeEvent = (DataChangeEvent) event;
+            return new SchemaAwareDataChangeEvent(
+                    dataChangeEvent, schemaHolder.getCurrentSchema(dataChangeEvent.tableId()));
+        } else {
+            throw new IllegalStateException(
+                    String.format(
+                            "Unrecognized event type \"%s\"", event.getClass().getCanonicalName()));
+        }
+    }
+
+    private static class SchemaHolder {
+        private final MapState<TableId, Schema> schemas;
+
+        public SchemaHolder(MapState<TableId, Schema> schemas) {
+            this.schemas = schemas;
+        }
+
+        public void applySchemaChange(SchemaChangeEvent schemaChangeEvent) throws Exception {
+            if (schemaChangeEvent instanceof CreateTableEvent) {
+                CreateTableEvent createTableEvent = (CreateTableEvent) schemaChangeEvent;
+                if (schemas.get(createTableEvent.tableId()) != null) {
+                    throw new IllegalStateException(
+                            String.format(
+                                    "Unable to apply CreateTableEvent to an existing schema for table \"%s\"",
+                                    createTableEvent.tableId()));
+                }
+                schemas.put(createTableEvent.tableId(), createTableEvent.getSchema());
+            } else if (schemaChangeEvent instanceof AlterColumnTypeEvent) {
+                handleAlterColumnTypeEvent((AlterColumnTypeEvent) schemaChangeEvent);
+            } else if (schemaChangeEvent instanceof RenameColumnEvent) {
+                handleRenameColumnEvent(((RenameColumnEvent) schemaChangeEvent));
+            } else if (schemaChangeEvent instanceof DropColumnEvent) {
+                handleDropColumnEvent(((DropColumnEvent) schemaChangeEvent));
+            } else if (schemaChangeEvent instanceof AddColumnEvent) {
+                handleAddColumnEvent(((AddColumnEvent) schemaChangeEvent));
+            } else {
+                throw new IllegalArgumentException(
```

Comment:
Unsupported Exception? Tell users that the framework does not support this kind of schema change, should support in framework firstly if they want to add more schema change event.